### PR TITLE
update mocha

### PR DESCRIPTION
**Pull request by akostadinov**
mocha 1.1.0 seems broken, see #3618

### CODE REVIEW

**Review by jlledom — COMMENTED** [review]
There are many tests failing. I'll approve when they pass.

**Review by jlledom — APPROVED** [review]
> Everybody would approve if tests passed. I thought you were a friend.

Oh, what the hell, approved.

**Comment by akostadinov** [comment]
@jlledom , a little bit more work now :grimacing:

File: Gemfile.lock [hunk 1]
```diff
@@ -1070,4 +1070,4 @@ DEPENDENCIES
   zip-zip
 
 BUNDLED WITH
-   2.2.25
+   2.3.25
```
Comment (jlledom) [inline Gemfile.lock:1073]: Better don't change this. We can upgrade bundler in another PR.

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -124,11 +124,33 @@
...
   end
 
+  config.after(:each) do |ex|
+    # I hope this would be easy to maintain in the long run, but it is optional.
+    # Just to remind anybody trying to use Mocha expectations in specs that this is not mandated.
+    has_expectations = Mocha::Mockery.instance.mocks.any? do |mock|
+      mock.__expectations__.to_a.any? do |expectation|
+        !expectation.instance_variable_get(:@cardinality).allowed_any_number_of_times?
+      end
+    end
```
Comment (jlledom) [inline spec/rails_helper.rb:134]: This seems to access Mocha internal variables and methods. If they change any of these in a next release, this will stop working. We'll have to be aware of this every time we want to update mocha, which I think we'll forget. I would look for another way to implement this check, or remove it if it's optional.

File: test/unit/logic/provider_upgrade_test.rb [hunk 3]
```diff
@@ -6,19 +6,21 @@ def setup
...
+    rule.stubs(:switches).returns(%i[finance multiple_applications branding require_cc_on_signup
       account_plans multiple_users groups]
     )
-    @power1M.plan_rule.stubs(:limits).returns(PlanRule::Limit.new(max_services: 1, max_users: 1))
-    @power1M.plan_rule.stubs(:rank).returns(10)
+    rule.stubs(:limits).returns(PlanRule::Limit.new(max_services: 1, max_users: 1))
+    rule.stubs(:rank).returns(10)
     @pro = FactoryBot.create(:published_plan, :system_name => 'pro3M', :issuer => service)
-    @pro.plan_rule.stubs(:switches).returns(%i[finance multiple_applications branding require_cc_on_signup account_plans
+    rule = @pro.send :plan_rule
```
Comment (jlledom) [inline test/unit/logic/provider_upgrade_test.rb:16]: why reusing the name for this variable? Why not use a different one? it would be less confusing.

File: test/unit/mailers/account_mailer_test.rb [hunk 4]
```diff
@@ -74,7 +74,7 @@ def test_support_entitlements_assigned
 
     def test_support_entitlements_trial_plan
       trial_plan = FactoryBot.create(:application_plan, issuer: @service)
-      trial_plan.plan_rule.metadata = {trial: true}
+      trial_plan.send(:plan_rule).metadata = {trial: true}
```
Comment (jlledom) [inline test/unit/mailers/account_mailer_test.rb:77]: why this change? Is it not equivalent?

**Review by jlledom — APPROVED** [review]

File: test/unit/logic/provider_upgrade_test.rb (hunk 3 above)
Comment (akostadinov) [reply test/unit/logic/provider_upgrade_test.rb:16]: Just saving in a local variable to avoid duplicate calls. Suggestions for naming?

Comment (jlledom) [reply test/unit/logic/provider_upgrade_test.rb:16]: `pro_rule` ?

File: Gemfile.lock (hunk 1 above)
Comment (akostadinov) [reply Gemfile.lock:1073]: mistake, will fix it, thanks!

File: spec/rails_helper.rb (hunk 2 above)
Comment (akostadinov) [reply spec/rails_helper.rb:134]: It is optional. If api changes, then test suite would fail. And we will know. So my line of thinking is that we can keep it until it becomes a problem and remove if we want later. Maybe comment not clear enough?

File: test/unit/mailers/account_mailer_test.rb (hunk 4 above)
Comment (akostadinov) [reply test/unit/mailers/account_mailer_test.rb:77]: mocha 2.x stubs private methods as private. Previously all was stubbed public. So we can't call directly anymore.

File: Gemfile.lock (hunk 1 above)
Comment (akostadinov) [reply Gemfile.lock]: fixed

File: test/unit/logic/provider_upgrade_test.rb (hunk 3 above)
Comment (akostadinov) [reply test/unit/logic/provider_upgrade_test.rb]: fixed

File: test/unit/mailers/account_mailer_test.rb (hunk 4 above)
Comment (jlledom) [reply test/unit/mailers/account_mailer_test.rb:77]: I wonder why ruby implements private methods if private protection is so easy to bypass.

**Review by jlledom — APPROVED** [review]

File: test/unit/mailers/account_mailer_test.rb (hunk 4 above)
Comment (akostadinov) [reply test/unit/mailers/account_mailer_test.rb:77]: It's to notify the user more than enforce restrictions. Still somehow useful.
